Custom type with multiple order_by clauses breaks pagination under asyncpg

Similar to #104 it appears that the when constructing the tuple for comparison sqlalchemy or sqlakeyset isn't aware of the custom type and is inserting a `::VARCHAR` cast instead of the custom type.

I am attempting to use a `UserDefinedType` to enable use of the [postgres ulid extension](https://github.com/pksunkara/pgx_ulid)

For single `order_by` clauses of/with the ulid id it works fine. However adding an additional `order_by` clause causes sqlakeyset to use a tuple for comparison and attempts to cast the custom type value to `::VARCHAR` instead of `::ulid`.

The error looks like:
```
E                   sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedFunctionError'>: operator does not exist: ulid < character varying
E                   HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
E                   [SQL: SELECT "user".id, "user".username 
E                   FROM "user" 
E                   WHERE ($1::VARCHAR, "user".id) < ("user".username, $2::VARCHAR) ORDER BY "user".username ASC, "user".id DESC 
E                    LIMIT $3::INTEGER]
E                   [parameters: ('Test1', '01J732P6YA8PS9TR7RMNF841QM', 2)]
E                   (Background on this error at: https://sqlalche.me/e/20/f405)

venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:784: ProgrammingError
```

It is entirely possible that I am holding it wrong or haven't implemented the `PgULID(UserDefinedType)` correctly :(

A test demonstrating the issue:

```python
class PgULID(UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "ulid"

    def bind_processor(self, dialect: Dialect):
        def process(value):
            if value:
                return str(value)

        return process

    def result_processor(self, dialect: Dialect, coltype: object):
        def process(value):
            if value:
                return ulid.from_str(value)

        return process


class TableBase(DeclarativeBase):
    type_annotation_map = {
        ulid.ULID: PgULID,
    }


class User(TableBase):
    __tablename__ = "user"
    id: Mapped[ulid.ULID] = mapped_column(primary_key=True)
    username: Mapped[str] = mapped_column()

    def __str__(self):
        return f"User({self.id}, {self.username})"

    def __repr__(self):
        return f"User({self.id}, {self.username})"


custom_bookmark_type(ulid.ULID, "ulid", ulid.from_str, str)

@pytest.mark.asyncio
async def test_ulid(engine, session_maker):

    async with engine.begin() as econn:
        await econn.execute(sa.text("CREATE EXTENSION IF NOT EXISTS ulid"))
        # conn.dialect.ischema_names["ulid"] = PgULID
        await econn.run_sync(TableBase.metadata.create_all)

    async with session_maker() as conn:
        await conn.execute(
            sa.insert(User).values([
                {"id": ulid.new(), "username": "Test1"},
                {"id": ulid.new(), "username": "Test2"},
            ])
        )
        query = sa.select(User)
        query = query.order_by(sa.asc(User.username))
        # commenting out this order_by works
        query = query.order_by(sa.desc(User.id))
        # E                   sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedFunctionError'>: operator does not exist: ulid < character varying
        # E                   HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
        # E                   [SQL: SELECT "user".id, "user".username
        # E                   FROM "user"
        # E                   WHERE ($1::VARCHAR, "user".id) < ("user".username, $2::VARCHAR) ORDER BY "user".username ASC, "user".id DESC
        # E                    LIMIT $3::INTEGER]
        # E                   [parameters: ('Test1', '01J732P6YA8PS9TR7RMNF841QM', 2)]
        # E                   (Background on this error at: https://sqlalche.me/e/20/f405)
        r = await conn.execute(query)
        users = []
        for index, row in enumerate(r.scalars()):
            print(index, row)
            users.append(row)

        page = await select_page(conn, query, per_page=1)
        assert [row[0] for row in page] == users[:1]
        assert page.paging.next is not None
        bookmark_next = page.paging.bookmark_next
        print(f"BOOKMARK: {bookmark_next}")

        page2 = await select_page(conn, query, per_page=1, page=bookmark_next)
        assert [row[0] for row in page2] == users[1:]
        assert page2.paging.has_next is False
```

Dockerfile to install the ulid extension into the postgres image
```dockerfile
FROM docker.io/library/postgres:15

ARG ULID_DEB=https://github.com/pksunkara/pgx_ulid/releases/download/v0.1.4/pgx_ulid-v0.1.4-pg15-amd64-linux-gnu.deb
WORKDIR /build
RUN /bin/true \
    && apt update \
    && apt install -y curl \
    && curl -L -o $PWD/pgx_ulid.deb $ULID_DEB \
    && apt install -y $PWD/pgx_ulid.deb \
    && apt-get clean
```

requirements.txt
```
sqlalchemy
ulid-py
pytest
pytest-asyncio
pytest-postgresql
psycopg[binary]
psycopg2-binary
sqlakeyset
asyncpg
```

--- a/sqlakeyset/paging.py
+++ b/sqlakeyset/paging.py
@@ -14,20 +14,21 @@ from typing import (
     Union,
     overload,
 )
-from typing_extensions import Literal  # to keep python 3.7 support
 
-from sqlalchemy import tuple_, and_, or_
+from sqlalchemy import and_, or_, tuple_
 from sqlalchemy.engine import Connection
 from sqlalchemy.engine.interfaces import Dialect
 from sqlalchemy.orm import Session
 from sqlalchemy.orm.query import Query
 from sqlalchemy.sql.expression import ColumnElement
 from sqlalchemy.sql.selectable import Select
+from typing_extensions import Literal  # to keep python 3.7 support
 
 from .columns import OC, MappedOrderColumn, find_order_key, parse_ob_clause
 from .results import Page, Paging, unserialize_bookmark
 from .serial import InvalidPage
 from .sqla import (
+    Row,
     core_coerce_row,
     core_result_type,
     get_bind,
@@ -37,14 +38,13 @@ from .sqla import (
     orm_query_keys,
     orm_result_type,
     orm_to_selectable,
-    Row,
 )
 from .types import Keyset, Marker, MarkerLike
-
 
 _TP = TypeVar("_TP", bound=Tuple[Any, ...])
 
 PER_PAGE_DEFAULT = 10
+
 
 def can_use_native_tuples(dialect: Dialect):
     # SQL drivers built-in to sqlalchemy that support native tuple comparison and seem
@@ -161,8 +161,7 @@ def prepare_paging(
     backwards: bool,
     orm: Literal[True],
     dialect: Dialect,
-) -> _PagingQuery:
-    ...
+) -> _PagingQuery: ...
 
 
 @overload
@@ -173,8 +172,7 @@ def prepare_paging(
     backwards: bool,
     orm: Literal[False],
     dialect: Dialect,
-) -> _PagingSelect:
-    ...
+) -> _PagingSelect: ...
 
 
 def prepare_paging(

--- a/sqlakeyset/paging.py
+++ b/sqlakeyset/paging.py
@@ -46,10 +46,17 @@ _TP = TypeVar("_TP", bound=Tuple[Any, ...])
 
 PER_PAGE_DEFAULT = 10
 
-# Dialects built-in to sqlalchemy that support native tuple comparison.
-# Other custom dialects may support this too, but we err on the side of
-# breaking less.
-SUPPORTS_NATIVE_TUPLE_COMPARISON = ("postgresql", "mysql", "sqlite")
+def can_use_native_tuples(dialect: Dialect):
+    # SQL drivers built-in to sqlalchemy that support native tuple comparison and seem
+    # to work fine with custom types in tuples. Other custom dialects may support this
+    # too, but we err on the side of breaking less.
+    if dialect.name in ("mysql", "sqlite"):
+        return True
+    if dialect.driver == "psycopg2":
+        # Other postgres drivers use render_bind_casts, which breaks our workaround for
+        # https://github.com/sqlalchemy/sqlalchemy/issues/8992.
+        return True
+    return False
 
 
 def compare_tuples(lesser: Sequence, greater: Sequence) -> ColumnElement[bool]:
@@ -91,9 +98,7 @@ def where_condition_for_page(
             "Page marker has different column count to query's order clause"
         )
 
-    native_tuple = (
-        len(place) > 1 and dialect.name.lower() in SUPPORTS_NATIVE_TUPLE_COMPARISON
-    )
+    native_tuple = len(place) > 1 and can_use_native_tuples(dialect)
 
     zipped = zip(ordering_columns, place)
     swapped = [

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,8 +1,8 @@
 import enum
+import uuid
 from datetime import timedelta
 from functools import partial
 from random import randrange
-import uuid
 
 import arrow
 import pytest
@@ -22,13 +22,12 @@ from sqlalchemy import select as _select
 from sqlalchemy.ext.hybrid import hybrid_property
 from sqlalchemy.orm import column_property, relationship
 from sqlalchemy.types import TypeDecorator
+from sqlalchemy_utils import ArrowType
 from sqlbag import S as _S
 from sqlbag import temporary_database
 
-from sqlakeyset.sqla import SQLA_VERSION
 from sqlakeyset import custom_bookmark_type
-
-from sqlalchemy_utils import ArrowType
+from sqlakeyset.sqla import SQLA_VERSION
 
 SQLA2 = SQLA_VERSION >= version.parse("1.4")
 if not SQLA2:
@@ -332,7 +331,13 @@ def _dburl(request):
 
     with temporary_database(request.param, host="localhost") as dburl:
         with S(dburl) as s:
-            Base.metadata.create_all(s.connection())
+            if request.param == "postgresql":
+                tables = None
+            else:
+                tables = [
+                    t for k, t in Base.metadata.tables.items() if not k.startswith("pg_only_")
+                ]
+            Base.metadata.create_all(s.connection(), tables=tables)
             s.add_all(data)
             s.execute(insert(Widget).values(widgets))
         yield dburl

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -335,7 +335,9 @@ def _dburl(request):
                 tables = None
             else:
                 tables = [
-                    t for k, t in Base.metadata.tables.items() if not k.startswith("pg_only_")
+                    t
+                    for k, t in Base.metadata.tables.items()
+                    if not k.startswith("pg_only_")
                 ]
             Base.metadata.create_all(s.connection(), tables=tables)
             s.add_all(data)

--- a/tests/test_async.py
+++ b/tests/test_async.py
@@ -1,15 +1,19 @@
-from contextlib import asynccontextmanager
 import json
 import re
+from contextlib import asynccontextmanager
 
 import pytest
 import pytest_asyncio
-from sqlalchemy.types import UserDefinedType
 from conftest import SQLA_VERSION, Author, Base, Book, Ticket
 from packaging import version
-from sqlalchemy import JSON, Column, Integer, String, desc, orm, select
+from sqlalchemy import Column, Integer, desc, orm, select
+from sqlalchemy.types import UserDefinedType
 
-from sqlakeyset.results import custom_bookmark_type, serialize_bookmark, unserialize_bookmark
+from sqlakeyset.results import (
+    custom_bookmark_type,
+    serialize_bookmark,
+    unserialize_bookmark,
+)
 
 if SQLA_VERSION < version.parse("1.4.0"):
     pytest.skip(
@@ -33,11 +37,13 @@ class StringifiedJSON(UserDefinedType):
     def bind_processor(self, dialect):
         def process(value):
             return json.dumps(value)
+
         return process
 
     def result_processor(self, dialect, coltype):
         def process(value):
             return value
+
         return process
 
     def get_col_spec(self, **kw):
@@ -50,7 +56,9 @@ class Jason(Base):
     id = Column(Integer, primary_key=True)
     content = Column(StringifiedJSON, nullable=False)
 
+
 custom_bookmark_type(dict, "D", json.loads, json.dumps)
+
 
 @asynccontextmanager
 async def _make_async_session(dburl):
@@ -65,18 +73,22 @@ async def _make_async_session(dburl):
 
     async with sessionmaker() as s:
         if pg:
-            s.add_all([
-                Jason(content={"hello": "world"}),
-                Jason(content={}),
-                Jason(content={"video": "games"}),
-                Jason(content={"potato": "salad"}),
-            ])
+            s.add_all(
+                [
+                    Jason(content={"hello": "world"}),
+                    Jason(content={}),
+                    Jason(content={"video": "games"}),
+                    Jason(content={"potato": "salad"}),
+                ]
+            )
         yield s
+
 
 @pytest_asyncio.fixture
 async def async_session(dburl):
     async with _make_async_session(dburl) as s:
         yield s
+
 
 @pytest_asyncio.fixture
 async def asyncpg_session(pg_only_dburl):
@@ -135,6 +147,7 @@ async def test_async_orm_query1(async_session):
 async def test_uuid(async_session):
     q = select(Ticket).order_by(Ticket.id)
     await check_paging_async(q, async_session)
+
 
 @pytest.mark.asyncio
 async def test_bind_processor(asyncpg_session):

--- a/tests/test_async.py
+++ b/tests/test_async.py
@@ -1,18 +1,15 @@
+from contextlib import asynccontextmanager
 import json
 import re
 
 import pytest
 import pytest_asyncio
+from sqlalchemy.types import UserDefinedType
 from conftest import SQLA_VERSION, Author, Base, Book, Ticket
 from packaging import version
-from sqlalchemy import Column, Integer, desc, orm, select
-from sqlalchemy.types import UserDefinedType
+from sqlalchemy import JSON, Column, Integer, String, desc, orm, select
 
-from sqlakeyset.results import (
-    custom_bookmark_type,
-    serialize_bookmark,
-    unserialize_bookmark,
-)
+from sqlakeyset.results import custom_bookmark_type, serialize_bookmark, unserialize_bookmark
 
 if SQLA_VERSION < version.parse("1.4.0"):
     pytest.skip(
@@ -30,37 +27,34 @@ ASYNC_PROTOS = {
 }
 
 
-class StringInt(UserDefinedType):
+class StringifiedJSON(UserDefinedType):
     cache_ok = True
 
     def bind_processor(self, dialect):
         def process(value):
-            return str(value)
-
+            return json.dumps(value)
         return process
 
     def result_processor(self, dialect, coltype):
         def process(value):
             return value
-
         return process
 
     def get_col_spec(self, **kw):
-        return "INT"
+        return "JSONB"
 
 
-class Wat(Base):
-    __tablename__ = "wat"
+class Jason(Base):
+    __tablename__ = "pg_only_jason"
 
     id = Column(Integer, primary_key=True)
-    num = Column(StringInt, nullable=False)
-
+    content = Column(StringifiedJSON, nullable=False)
 
 custom_bookmark_type(dict, "D", json.loads, json.dumps)
 
-
-@pytest_asyncio.fixture
-async def async_session(dburl):
+@asynccontextmanager
+async def _make_async_session(dburl):
+    pg = dburl.startswith("postgres")
     for k, v in ASYNC_PROTOS.items():
         dburl = re.sub("^" + k, v, dburl)
     engine = asa.create_async_engine(dburl, future=True)
@@ -70,7 +64,23 @@ async def async_session(dburl):
         sessionmaker = orm.sessionmaker(engine, class_=asa.AsyncSession)
 
     async with sessionmaker() as s:
-        s.add_all([Wat(num=i % 3) for i in range(20)])
+        if pg:
+            s.add_all([
+                Jason(content={"hello": "world"}),
+                Jason(content={}),
+                Jason(content={"video": "games"}),
+                Jason(content={"potato": "salad"}),
+            ])
+        yield s
+
+@pytest_asyncio.fixture
+async def async_session(dburl):
+    async with _make_async_session(dburl) as s:
+        yield s
+
+@pytest_asyncio.fixture
+async def asyncpg_session(pg_only_dburl):
+    async with _make_async_session(pg_only_dburl) as s:
         yield s
 
 
@@ -126,8 +136,7 @@ async def test_uuid(async_session):
     q = select(Ticket).order_by(Ticket.id)
     await check_paging_async(q, async_session)
 
-
 @pytest.mark.asyncio
-async def test_bind_processor(async_session):
-    q = select(Wat).order_by(Wat.num, Wat.id)
-    await check_paging_async(q, async_session)
+async def test_bind_processor(asyncpg_session):
+    q = select(Jason).order_by(Jason.content, Jason.id)
+    await check_paging_async(q, asyncpg_session)

--- a/tests/test_async.py
+++ b/tests/test_async.py
@@ -1,12 +1,18 @@
+import json
 import re
 
 import pytest
 import pytest_asyncio
-from conftest import SQLA_VERSION, Author, Book
+from conftest import SQLA_VERSION, Author, Base, Book, Ticket
 from packaging import version
-from sqlalchemy import desc, orm, select
+from sqlalchemy import Column, Integer, desc, orm, select
+from sqlalchemy.types import UserDefinedType
 
-from sqlakeyset.results import serialize_bookmark, unserialize_bookmark
+from sqlakeyset.results import (
+    custom_bookmark_type,
+    serialize_bookmark,
+    unserialize_bookmark,
+)
 
 if SQLA_VERSION < version.parse("1.4.0"):
     pytest.skip(
@@ -24,6 +30,35 @@ ASYNC_PROTOS = {
 }
 
 
+class StringInt(UserDefinedType):
+    cache_ok = True
+
+    def bind_processor(self, dialect):
+        def process(value):
+            return str(value)
+
+        return process
+
+    def result_processor(self, dialect, coltype):
+        def process(value):
+            return value
+
+        return process
+
+    def get_col_spec(self, **kw):
+        return "INT"
+
+
+class Wat(Base):
+    __tablename__ = "wat"
+
+    id = Column(Integer, primary_key=True)
+    num = Column(StringInt, nullable=False)
+
+
+custom_bookmark_type(dict, "D", json.loads, json.dumps)
+
+
 @pytest_asyncio.fixture
 async def async_session(dburl):
     for k, v in ASYNC_PROTOS.items():
@@ -35,6 +70,7 @@ async def async_session(dburl):
         sessionmaker = orm.sessionmaker(engine, class_=asa.AsyncSession)
 
     async with sessionmaker() as s:
+        s.add_all([Wat(num=i % 3) for i in range(20)])
         yield s
 
 
@@ -82,4 +118,16 @@ async def check_paging_async(selectable, s):
 async def test_async_orm_query1(async_session):
     spec = [desc(Book.b), Book.d, Book.id]
     q = select(Book, Author, Book.id).outerjoin(Author).order_by(*spec)
+    await check_paging_async(q, async_session)
+
+
+@pytest.mark.asyncio
+async def test_uuid(async_session):
+    q = select(Ticket).order_by(Ticket.id)
+    await check_paging_async(q, async_session)
+
+
+@pytest.mark.asyncio
+async def test_bind_processor(async_session):
+    q = select(Wat).order_by(Wat.num, Wat.id)
     await check_paging_async(q, async_session)

